packages/backend-app-api/src/services/implementations/discovery/MultipleBackendHostDiscovery.ts

```diff
@@ -0,0 +1,166 @@
+  RootFeatureRegistryService,
+} from '@backstage/backend-plugin-api';
+import { Config } from '@backstage/config';
+import fetch from 'node-fetch';
+import { NotFoundError } from '@backstage/errors';
+
+interface DiscoveryUrl {
+  internal: string;
+  external: string;
+}
+
+export type PluginRegistrations = Record<string, DiscoveryUrl>;
+
+export class MultipleBackendHostDiscovery implements DiscoveryService {
+  #gatewayUrl: string;
+  #instanceUrl: string;
+
+  #isGateway: boolean;
+  #isInitialized = false;
+
+  #discovery: DiscoveryService;
+  #rootFeatureRegistry: RootFeatureRegistryService;
+
+  // A map of plugin to URLs.
+  #plugins: PluginRegistrations = {};
+  // A map of instance URL to plugin.
+  #instancePlugins: Record<string, Set<string>> = {};
+
+  static fromConfig(
+    config: Config,
+    options: {
+      rootFeatureRegistry: RootFeatureRegistryService;
+      basePath?: string;
+    },
+  ) {
+    return new MultipleBackendHostDiscovery({
+      instanceUrl: config.getString('backend.baseUrl'),
+      gatewayUrl: config.getOptionalString('discovery.gatewayUrl'),
+      rootFeatureRegistry: options.rootFeatureRegistry,
+      discovery: HostDiscovery.fromConfig(config),
+    });
+  }
+
+  constructor(options: {
+    instanceUrl: string;
+    gatewayUrl?: string;
+    rootFeatureRegistry: RootFeatureRegistryService;
+    discovery: DiscoveryService;
+  }) {
+    this.#gatewayUrl = options.gatewayUrl || options.instanceUrl;
+    this.#instanceUrl = options.instanceUrl;
+    this.#isGateway = !options.gatewayUrl;
+    this.#discovery = options.discovery;
+    this.#rootFeatureRegistry = options.rootFeatureRegistry;
+  }
+
+  async initialize() {
+    if (this.#isInitialized) {
+      throw new Error('Can not initialize twice.');
+    }
+    const features = await this.#rootFeatureRegistry.getFeatures();
+    const pluginIds = features
+      .filter(e => e.type === 'plugin')
+      .map(e => e.pluginId);
+    const plugins: PluginRegistrations = {};
+    for (const pluginId of pluginIds) {
+      plugins[pluginId] = {
+        internal: await this.#discovery.getBaseUrl(pluginId),
+        external: await this.#discovery.getExternalBaseUrl(pluginId),
+      };
+    }
+    this.addPlugins(this.#instanceUrl, plugins);
+  }
+
+  addPlugins(instanceUrl: string, plugins: PluginRegistrations) {
+    for (const [pluginId, urls] of Object.entries(plugins)) {
+      this.#plugins[pluginId] = urls;
+      if (!this.#instancePlugins[instanceUrl]) {
+        this.#instancePlugins[instanceUrl] = new Set();
+      }
+      this.#instancePlugins[instanceUrl].add(pluginId);
+    }
+  }
+
+  get plugins() {
+    return this.#plugins;
+  }
+
+  get instancePlugins() {
+    return this.#instancePlugins;
+  }
+
+  get isGateway() {
+    return this.#isGateway;
+  }
+
+  async getUrl(pluginId: string, key: 'external' | 'internal') {
+    // Because of how the services are initialized, lazy load the features when we first need them.
```

Comment:
`proxy-backend` is the main issue here. it's a service on the current instance and when it starts up it wants its external URL. Ideally, we would initialize this class when the lifecycle hook `onStartupFinished`, but we need to lazy load the startup as the plugin initialization happens after registration.

Reply:
On the other side, we need to lazy load here because the `rootFeatureRegistry` isn't full until after the backend starts, so we can't do this in the constructor.